### PR TITLE
OpenAPI skills MVP

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
### Motivation and Context

The change lay foundation for OpenApi skills support by introducing a set of classes for importing, parsing OpenApi documents and sending HTTP requests to the resources described in them.

### Description

The PR adds the following classes/methods:
- ImportOpenApiFromDirectoryExtension – discovers OpenAPI document in a folder, creates SKFunction per REST operation in the document, and registers the functions in the Kernel. 
- RegisterOpenApiSkill - reates SKFunction per REST operation in the document, and registers the functions in the Kernel. 
- OpenApiDocumentParser – iterates over the collection of paths in OpenAPI document and creates an instance of RestOperation model class for each of them. 
- RestOperationRunner – accepts an instance of the RestOperation class and collection of arguments. Gets all required components (url, body, headers, …) for HTTP request and sends the request. 
- A few other model, exception, and helper classes.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Extensions/KernelOpenApiExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        {
+            try
+            {
+                kernel.Log.LogTrace("Registering Rest function {0}.{1}.", skillName, operation.Id);
+                skill[skillName] = kernel.RegisterRestFunction(skillName, operation);
+            }
+            catch (Exception ex) when (!ex.IsCriticalException())
+            {
+                //Logging the exception and keep registering other Rest functions
+                kernel.Log.LogWarning(ex, "Something went wrong while rendering the Rest function. Function: {0}.{1}. Error: {2}", skillName, operation.Id, ex.Message);
```
Comment (adrianwyatt) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Extensions/KernelOpenApiExtensions.cs:48]: Why not throw?

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameterType.cs
Comment (umeshma) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameterType.cs:1]: There are already part of the Microsoft.OpenApi SDK. All defined here. 
Microsoft.OpenApi.Models.ParameterLocation

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Extensions/KernelOpenApiExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+using RestSkills.Authentication;
+
+namespace Microsoft.SemanticKernel.Skills.OpenAPI.Extensions;
+
+/// <summary>
+/// Class for extensions methods for IKernel interface.
+/// </summary>
+public static class KernelOpenApiExtensions
+{
+    public static IDictionary<string, ISKFunction> RegisterOpenApiSkill(this IKernel kernel, Stream documentStream, string skillName)
```
Comment (shawncal) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Extensions/KernelOpenApiExtensions.cs:26]: We'll want a overload of this that loads directly from a json url, and one from a local file json path (maybe these are the same... both string paths)

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestOperation.cs
Comment (umeshma) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestOperation.cs:1]: There is already a OpenApiOperation object.
Microsoft.OpenApi.Models.OpenApiOperation

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/OpenApi/OpenApiDocumentParser.cs
Comment (umeshma) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/OpenApi/OpenApiDocumentParser.cs:1]: Why duplicate all the parameter information into internal objects? 
I can see why you may want a REST operation, but why duplicate all the parameters and options? Just hold onto all the metadata types already defined in Microsoft.OpenAPI - keep a reference to them.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Skills/LocalOpenApiSkillsProvider.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    /// Azure KeyValut skill name.
+    /// </summary>
+    public const string AzureKeyVault = "AzureKeyVaultSkill";
+
+    /// <summary>
+    /// Loads OpenApi document from assembly resource.
+    /// </summary>
+    /// <param name="skillName">The skill name.</param>
+    /// <returns>The OpenApi document resource stream.</returns>
+    public static Stream LoadFroResource(string skillName)
```
Comment (shawncal) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Skills/LocalOpenApiSkillsProvider.cs:24]:
```suggestion
    public static Stream LoadFromResource(string skillName)
```

File: samples/dotnet/kernel-syntax-examples/Example16_OpenApiSkill.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        Console.WriteLine("======== Planning - Create and Execute Azure Plan ========");
+        var kernel = InitializeKernelAndPlanner(out var planner);
+
+        //Use OpenApi skill from folder
+        //string folder = RepoFiles.SampleSkillsPath();
+        //kernel.ImportOpenApiSkillFromDirectory(folder, "AzureKeyVaultSkill");
+
+        //Use OpenApi skill from Skills.OpenAPI package
+        kernel.RegisterOpenApiSkill(LocalOpenApiSkillsProvider.LoadFroResource(LocalOpenApiSkillsProvider.AzureKeyVault), LocalOpenApiSkillsProvider.AzureKeyVault);
```
Comment (shawncal) [inline samples/dotnet/kernel-syntax-examples/Example16_OpenApiSkill.cs:33]:
```suggestion
        kernel.RegisterOpenApiSkill(LocalOpenApiSkillsProvider.LoadFromResource(LocalOpenApiSkillsProvider.AzureKeyVault), LocalOpenApiSkillsProvider.AzureKeyVault);
```

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Rest/RestOperationRunner.cs
Comment (umeshma) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Rest/RestOperationRunner.cs:1]: Where will retry etc. get applied for the HttpClient?

File: samples/dotnet/kernel-syntax-examples/Example16_OpenApiSkill.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+        var plan = await kernel.RunAsync("Load 'test-secret' from Azure KeyValut available at https://dev-tests.vault.azure.net using GetSecret function.", planner["CreatePlan"]);
```
Comment (adrianwyatt) [inline samples/dotnet/kernel-syntax-examples/Example16_OpenApiSkill.cs:35]: In the future I would recommend not using the planner for this particular example, so as to keep the concept count limited to what we are trying to show.

**Review by adrianwyatt — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Rest/RestOperationRunner.cs
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Rest/RestOperationRunner.cs:1]: Good callout. This is going into a feature branch for iteration and we'll be taking a look at robustness before merging into main.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameter.cs
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameter.cs:1]: @SergeyMenshykh  - Umesh has a good point. Though, to keep other work moving forward, I'm happy to bring this into the feature branch as-is if we look at replacing these models with Microsoft.OpenApi. Unless you have other reasons to roll our own.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/OpenApi/OpenApiDocumentParser.cs
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/OpenApi/OpenApiDocumentParser.cs:1]: We will evaluate this change in a (near) future change.

**Review by adrianwyatt — APPROVED** [review]

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameterType.cs
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameterType.cs:1]: We will evaluate this change in a (near) future change.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameter.cs
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestParameter.cs:1]: We will evaluate this change in a (near) future change.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestOperation.cs
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestOperation.cs:1]: We will evaluate this change in a (near) future change.

**Comment by shawncal** [comment]
@SergeyMenshykh and @adrianwyatt Now that this is in the feature branch (which is great!) I've un-resolved all the unresolved comments so that they can be addressed in a pre-merge PR. Please make sure to note them (or they'll likely all come up again when merging to main).

**Pull request merged**